### DISCUSSION TITLE
How to trigger configfile generation?

### DISCUSSION THREAD

**Discussion by davidchisnall** [xmake-io/xmake#2846] [Q&A]
I am trying to construct a reduced test case from a larger system where a file generated with `add_configfiles` is generated only during the initial configure step and never during the build (even if the source file is modified).  I started with this:

```
$ cat xmake.lua
target("test")
        set_kind("binary")
        add_files("a.c")
        add_configfiles("foo.in")
        set_configvar("a", "b")
$ cat a.c
int main() {}
$ cat foo.in
${a}
```

But this doesn't generate the output file at all.  What do I need to add to make this work?

**Comment by waruqi** [comment]
try pass `-c` to `xmake f -c` to regenerate it.

**Comment by waruqi** [comment]
try pass `-c` to `xmake f -c` to regenerate it.

**Reply by davidchisnall** [reply]
Thanks.  That works in both the simple and complex cases.  In the more complex case, just passing `xmake f -c` makes it forget other config parameters that I previously passed on the command line.  What should I be doing to make them persistent?

**Reply by waruqi** [reply]
you need pass full configuration arguments to `xmake f/config` each time. like configure/cmake

```console
xmake f -c --xx --yy
```

**Reply by davidchisnall** [reply]
CMake remembers the config (`cmake .` reruns the configuration pass, but remembers everything from the previous run.  Configuration settings are stored in `CMakeCache.txt`).  Autoconf scripts are meant to remember the arguments so that a zero-argument invocation doesn't change anything, but that logic is normally buggy (I've only seen a couple of projects where it worked reliably).

**Reply by waruqi** [reply]
`xmake f/config` without `-c` will remember all too. but `-c` argument will ignore all configuration cache to force regenerate all.

**Reply by davidchisnall** [reply]
I see.  Is there a way of forcing it to run all of the configuration steps with the current config, without doing `xmake f --export=tmp ; xmake f -c --import=tmp`?

**Reply by waruqi** [reply]
modify your xmake.lua file, it will do it.

**Reply by davidchisnall** [reply]
What is the hook that I need to implement?

**Reply by waruqi** [reply]
If nothing has changed in xmake.lua, you should not bother to re-detect and update the configuration. If you want to force a regeneration of all configurations, only add -c and all full parameters.

Of course, there are some improvements that could be made to configfiles, such as automatically regenerating configfiles if they have been changed.

**Reply by davidchisnall** [reply]
Regenerating the config file is the key part.  It's very surprising as a user that changing the .in file does not cause the output to be regenerated.

**Reply by waruqi** [reply]
I have improve it, you can try it again. https://github.com/xmake-io/xmake/pull/2854

```
xmake update -s github:xmake-io/xmake#config
```

**Reply by waruqi** [reply]
xmake update -s dev

**Reply by davidchisnall** [reply]
Thanks.  This seems to generate the config file every time the build is run.  This is fine for our use case (where we need it only in the final link) but will be painful for `.h` files that are included in something early in the build, because now large numbers of things will be rebuilt every build.  It would be nice if the config files were rebuilt only if they changed.

**Reply by davidchisnall** [reply]
Or does it only install the new file if it's changed?

**Reply by waruqi** [reply]
> It would be nice if the config files were rebuilt only if they changed.

It should be like that now, I tested it.

It will generate it when it is changed.

**Reply by waruqi** [reply]
> Or does it only install the new file if it's changed?

yes, now

**Reply by waruqi** [reply]
I have improved it on dev, it should work.
